### ISSUE TITLE
Samples insertion is super slow with SQLite

### ISSUE THREAD

**Issue by mkalam-alami** [alakajam-team/alakajam#220]
I had to lower sample user counts to 5 get a decent start time. Maybe transaction support can help.

**Comment by ttencate** [comment]
Development server startup without an existing database performs 3222 queries (counted using `DEBUG=knex:query node server.js`). Each of the 9x10 = 90 vote insertions (theoretically -- due to overlap it's only 73) seems to be taking around 40 queries, for a total of 2829. So that's where the bulk of the time is going.

This isn't surprising, given that we're doing an `O(n³)` algorithm: `n` people, voting on `n` entries, and upon each vote recounting all `O(n)` previous votes. I tried to suppress the recount, which cuts total server startup time down from 27 to 13 seconds. But this not a proper solution.

The proper solution would be transactions, as you already suggested. I tried to do a quick performance test but ran into issues like #265 again.

In fact, transactions should be used much more in general, for safer and saner code, with speedup as only a side benefit. But that means any `service` function that accesses the database will have to have the transaction passed in. Rather tedious!

For creating/running the transaction, we can probably write a small Express middleware, so every HTTP request gets its own transaction. This technically makes sense even if they just do `SELECTS`s, so each request gets a consistent view of the world, although I'm not sure what it would mean for performance.

**Comment by mkalam-alami** [comment]
> In fact, transactions should be used much more in general, for safer and saner code, with speedup as only a side benefit.

Agreed, in the current state we do have some places (especially when saving entries) that can lead to inconsistent DB if something fails. Usually that means a broken entry, which can at worst fixed by deleting the entry so it's usually not too damaging... But we might have nastier things waiting for their time to come :P

I wouldn't mind passing around transactions, though maybe it does need some Bookshelf integration tricks to prevent the code from getting bloated.

In a lot of cases it would probably fit to have a per-HTTP-request transaction (if only the GET ones?). Now design-wise it feels better to me to leave the responsability of transactions at the services layers: if service functions are not atomic, it's because they've been badly designed and/or we need a higher layer of services.

> I tried to do a quick performance test but ran into issues like #265 again.

Argh SQLite really is getting in the way. Opening #269 to find a once-and-for-all solution.

**Comment by ttencate** [comment]
> Now design-wise it feels better to me to leave the responsability of transactions at the services layers: if service functions are not atomic, it's because they've been badly designed and/or we need a higher layer of services.

Sounds fair, but it does mean that SQLite samples insertion will remain super slow ;) But maybe we can add an optional `transaction` argument to those particular service functions where it is needed. Or add functions for bulk operations.

**Comment by ttencate** [comment]
Just stumbled into https://www.sqlite.org/wal.html which might be a good stopgap solution if this is still bothering you. I didn't read it in detail but it looks like you can just send the following raw SQL command after connecting to SQLite:

    PRAGMA journal_mode=WAL;

**Comment by ttencate** [comment]
[Also worth trying](https://www.sqlite.org/pragma.html#pragma_synchronous):

    PRAGMA synchronous=OFF;
